### ISSUE TITLE
Implement `forEachItem` for JSON object

### ISSUE THREAD

**Issue by tfmorris** [OpenRefine/OpenRefine#3147]
Users have ongoing complaints about the difficult in parsing  JSON which has arbitrary structure, most often in the form of JSON API responses which were designed to return objects keyed by arbitrary identifiers rather than just returning array as the vast more of APIs do.

Examples provided in #2186 include:
* Wikimedia Commons -> https://commons.wikimedia.org/w/api.php?action=query&titles=File:Google%202015%20logo.svg&prop=imageinfo&iiprop=url&format=json
* MySociety MapIt -> https://mapit.mysociety.org/point/4326/0.1293497,51.5464828

And the issue has been discussed both on the OpenRefine mailing list and StackOverflow:
* https://groups.google.com/d/msg/openrefine/O72bg8GZkj8/vBoEzmGxBwAJ
* https://stackoverflow.com/questions/10782737/google-refine-iterate-over-a-json-dictionary

### Proposed solution

Similar to the `forEachIndex` function that we have for arrays, we would implement a `forEachItem` iterator control which would accept a JSON node object, two variable names (one for key and one for value), and a GREL expression. It would iterate over each item in the JSON object, assigning the string value of the key to the key variable and the value for the item to the value variable and then execute the GREL expression.

The control would **only** accept a JSON object to iterate over and would explicitly **not** support iterating over objects that implement the `HasFields` interface.

### Alternatives considered
* extend `forEach()` to iterate over JSON object items. This wouldn't provide access to the field name / key, but would probably satisfy the vast majority of cases since the key is usually included in the child object as an id.
* implement iteration for every object that implements HasFields. Since OpenRefine `HasFields` objects typically have a small fixed set of fields, this seems to have limit value.

**Comment by thadguidry** [comment]
@tfmorris I'm trying to wrap my head around your solution and how the function would actually look or work with a sample GREL expression and some sample JSON?  Can you provide that also please? It's probably just me.  Hmm, this "feels" like something already capable with the Filter syntax (`in`, `subsetof`) in JSONPath or JSONPath-Plus?

JSONPath
To retrieve all books that have the isbn property

Use the following expression:
`$..book[?(@.isbn)]`

maybe I'm wrong and there's nothing within the JSONPath bag of syntax that is close to your forEachItem, but I'll wait for your example GREL and JSON sample.

**Comment by tfmorris** [comment]
>  I'm trying to wrap my head around your solution and how the function would actually look or work with a sample GREL expression and some sample JSON? 

Look at [forEachIndex](https://github.com/OpenRefine/OpenRefine/wiki/GREL-Controls#foreachindexexpression-a-variable-i-variable-v-expression-e) and replace "variable i" for the index value with "variable k" for the key value. Everything else stays the same.

`forEachItem('{"key1": "vala", "key2": "valb"}'.parseJson(), key, value, key[-1]+value[-1])` returns  `["1a", "2b"]`

In every example I've seen, the keys are redundantly encoded in the value object, so I've implemented the simple case first to see if that's good enough.

**Comment by thadguidry** [comment]
@tfmorris With this proposal, could it then help to implement a new option on the Split multi-value cells?
* by Json objects in Json array

or perhaps we could make it just
* by objects in an array

![image](https://user-images.githubusercontent.com/986438/98827462-854bcc00-23fc-11eb-88b1-d7276edc1eff.png)

```
"DTName","SDTCode","SDTName","JSON"
"BIRBHUM","'02249","MURARAI -I","[{""name1"":""Birnagar"",""vid"":""2008000101303000""},{""name1"":""Birnagar"",""vid"":""2010000501792500""}]"
```

**Comment by tfmorris** [comment]
That sounds like a UI thing while this issue proposes a new GREL function (if #3149/#3150 don't satisfy the bulk of the need, which I suspect they might).

Split multi-valued cells could conceivably be extended to accept any GREL expression that generates an array, but it's not clear how many users would need that much power -- and you could compose that by doing `separator.join(array_generator())` and then splitting on your separator.

**Comment by thadguidry** [comment]
@tfmorris Ah, I missed your PR, let me try the recent snapshot version that includes it.

**Comment by satyamksharma** [comment]
Hi there, I would like to resolve this issue, please assign me!

**Comment by wetneb** [comment]
@satyamksharma I realize you are assigned to two good first issues: this one and #5656. I would prefer if you were only assigned to one: we are running short of good first issues and other applicants want to make contributions to the project. Which of those two issues are you currently working on? I would unassign you from the other one. Thank you!

**Comment by satyamksharma** [comment]
@wetneb I appreciate your concern and please go ahead but unassign the other one not this one. Thank you

**Comment by tfmorris** [comment]
So instead of creating the `forEachItem` control that I originally proposed, I've instead drafted a pull request extending `forEachIndex` to iterate over key/value pairs in addition to array index/value pairs. What do people think of that alternative approach?

**Comment by ostephens** [comment]
> So instead of creating the forEachItem control that I originally proposed, I've instead drafted a pull request extending forEachIndex to iterate over key/value pairs in addition to array index/value pairs. What do people think of that alternative approach?

Been trying to think of where this would be a problem, and the only scenario I've come up with is if the user doesn't know ahead of time whether they are dealing with an array or an object. At the moment obviously the array would work, and the object should give an error[1]. This, in theory, allows the user to not worry about the values that aren't arrays when applying `forEachIndex` - they can either store the error (although see note below) or blank, or the current value.

However, since the likelihood of having data where the same column contains sometimes an array and sometimes and object seems slim AND the user can use `type()` to check what's going in if they need to, I don't think this is really an issue.

So in summary - happy for `forEachIndex` to be extended as proposed



1. Note: it looks like `forEachIndex` currently returns a `string` rather than `error` if it is given something other than an array which is odd

**Comment by tfmorris** [comment]
@thadguidry No existing functionality is being proposed for removal.

@ostephens re:
> Note: it looks like forEachIndex currently returns a string rather than error if it is given something other than an array which is odd

It seems irregular, but it's how the public `Control` interface is defined:

https://github.com/OpenRefine/OpenRefine/blob/c457bc44b9f8a530dccc7ad81c95598824bb6ada/main/src/com/google/refine/grel/Control.java#L51

so changing it would require changing it for all controls (ie `for()`, `if()`, etc), not just `forEachIndex()`, plus the only code that uses this is the GREL Parser, which immediately converts it to an exception:

https://github.com/OpenRefine/OpenRefine/blob/c457bc44b9f8a530dccc7ad81c95598824bb6ada/main/src/com/google/refine/grel/Parser.java#L219-L221

**Comment by ostephens** [comment]
> so changing it would require changing it for all controls (ie for(), if(), etc), not just forEachIndex(), plus the only code that uses this is the GREL Parser, which immediately converts it to an exception:

Ah thanks @tfmorris  - I think the difference I was seeing (in 3.7.4) you've corrected with the PR linked her - the code was:

`return ControlEvalError.foreach_index();`

and is now

`return new EvalError(ControlEvalError.foreach_index());`

which brings it in line with the other controls

**Comment by tfmorris** [comment]
Oh, I was looking at the wrong code. You're talking about this:
https://github.com/OpenRefine/OpenRefine/pull/5997/files#diff-3e1b7f326e39cafe48a94652d4e0816d22c6b530878eab2d4823dc73d16a6f5cR79-R80

That issue was apparently introduced during the backend localization a year ago and we didn't have any test coverage for it. Although I improved the tests, it looks like that line still isn't covered. :(
https://coveralls.io/jobs/126170057/source_files/12525500499#L80
